Don't include repo on push if branch has upstream

rchl opened the pull request:
When the branch is tracking upstream branch there should
be no need to explicitly state the origin name. Especially as
that might not be the one that matches tracking information.

Resolves #608

rchl commented:
I can have a look at failing tests and add new one if you think it makes sense.

webpro commented:
Thanks @rchl! Looks good at first glance, but please also check and fix the tests. Would be much appreciated! And might even surface a bug.

rchl commented:
After looking more into it, I feel like there is no perfect way of implementing that **without** breaking backwards compatibility.

I think ideal solution would be to:
 - set default value of `git.pushRepo` to an empty string
and then:
 - if `git.pushRepo` is set by the user, use that
 - if `git.pushRepo` has default value:
   a) if branch is tracking upstream - push to that
   b) if branch is not tracking upstream - push to `origin`

With current logic where `git.pushRepo` is set to `origin` by default, it's impossible to know if that is just the default value or user really wants to force the `origin` repo`.

rchl commented:
Before I go further with it, I need to solve a mystery.
One of the tests is failing for me without me even doing any changes:
```
git › should push to remote name (not "origin")

  /private/var/folders/d8/c7ngbjld0px_b42v6xy_d4s00000gn/T/release-it-5596D7fjiaEgt5o0/lib/shell.js:49

  Rejected promise returned by test. Reason:

  Error {
    message: `fatal: You are pushing to remote 'upstream', which is not the upstream of␊
    your current branch 'master', without telling me what to push␊
    to update which remote branch.␊
    `,
  }
```
The error makes sense but I don't get why it's not triggering on Jenkins. Does Jenkins has some different global git settings that make this not trigger?
I've tried locally with both git `2.25.0` and `2.21.1 (Apple Git-122.3)` and both trigger error while on Jenkins `2.24.0` is used and it doesn't trigger this error...

rchl commented:
Fixed tests and updated logic.

This change could be considered breaking since there is one case where behavior will differ.
When:
 - a branch is tracking upstream
 - `git.pushRepo` is at default `origin` value

Previously that would trigger `git push origin` thus always push to origin even if tracking information were pointing to another remote.
With those changes, in the same situation, `git push` will be triggered and will push to upstream branch.

I think current behavior makes more sense but there is a very tiny chance someone might be affected by this change.

webpro commented:
Yeah, I've had similar challenges with `npm.tag: latest` and such (when setting a default value in release-it, unable to see if the user explicitly has set this, or not at all).

What might be interesting here is that I'm pretty far with work on [v13](https://github.com/release-it/release-it/commits/v13) (mainly about updating dependencies and dropping support for Node v8), so perhaps it makes sense to merge your ideal solution (I agree!) to this branch.

We could still merge your current work into v12, but I prefer not to introduce this potentially breaking change (basically the same reasoning I bumped to https://github.com/release-it/release-it/blob/master/CHANGELOG.md#v12).

Would you be OK with you implementing your ideal solution, merge into v13, and I'll release v13 happen soon?

rchl commented:
Sure :)

rchl commented:
I'm fine with skipping this in v12.
